app/service/sorting_test.go

```diff
@@ -0,0 +1,171 @@
+		{name: "repeated field",
+			args: args{
+				urlParameters: "?sort=name,name",
+				acceptedFields: map[string]*FieldSortingParams{
+					"name": {ColumnName: "sName", FieldType: "string"},
+					"id":   {ColumnName: "ID", FieldType: "int64"},
+				},
+				defaultRules: "-name,id",
+			},
+			wantAPIError: ErrInvalidRequest(errors.New(`a field cannot be a sorting parameter more than once: "name"`))},
+		{name: "unknown field",
+			args: args{
+				urlParameters: "?sort=class",
+				acceptedFields: map[string]*FieldSortingParams{
+					"name": {ColumnName: "sName", FieldType: "string"},
+					"id":   {ColumnName: "ID", FieldType: "int64"},
+				},
+				defaultRules: "-name,id",
+			},
+			wantAPIError: ErrInvalidRequest(errors.New(`unknown field in sorting parameters: "class"`))},
+		{name: "add id field",
+			args: args{
+				urlParameters: "",
+				acceptedFields: map[string]*FieldSortingParams{
+					"name": {ColumnName: "sName", FieldType: "string"},
+				},
+				defaultRules: "-name",
+			},
+			wantSQL:          "SELECT ID FROM `users` ORDER BY sName DESC, ID ASC",
+			wantSQLArguments: nil,
+			wantAPIError:     NoError},
+		{name: "no rules (adds id)",
+			args: args{
+				urlParameters: "",
+				acceptedFields: map[string]*FieldSortingParams{
+					"name": {ColumnName: "sName", FieldType: "string"},
+					"id":   {ColumnName: "ID", FieldType: "int64"},
+				},
+			},
+			wantSQL:      "SELECT ID FROM `users` ORDER BY ID ASC",
+			wantAPIError: NoError},
+		{name: "sorting + paging",
+			args: args{
+				urlParameters: "?from.id=1&from.name=Joe&from.flag=1",
+				acceptedFields: map[string]*FieldSortingParams{
+					"name": {ColumnName: "sName", FieldType: "string"},
+					"id":   {ColumnName: "ID", FieldType: "int64"},
+					"flag": {ColumnName: "bFlag", FieldType: "bool"},
+				},
+				defaultRules: "-name,id,flag",
+			},
+			wantSQL:          "SELECT ID FROM `users` WHERE ((sName < ?) OR (sName = ? AND ID > ?) OR (sName = ? AND ID = ? AND bFlag > ?)) ORDER BY sName DESC, ID ASC",
+			wantSQLArguments: []driver.Value{"Joe", "Joe", 1, "Joe", 1, true},
+			wantAPIError:     NoError},
+		{name: "wrong value in from.id field",
+			args: args{
+				urlParameters: "?from.id=abc&from.name=Joe",
+				acceptedFields: map[string]*FieldSortingParams{
+					"name": {ColumnName: "sName", FieldType: "string"},
+					"id":   {ColumnName: "ID", FieldType: "int64"},
+				},
+				defaultRules: "-name,id",
+			},
+			wantAPIError: ErrInvalidRequest(errors.New(`wrong value for from.id (should be int64)`))},
+		{name: "one of the from. fields is skipped",
+			args: args{
+				urlParameters: "?from.id=2",
+				acceptedFields: map[string]*FieldSortingParams{
+					"name": {ColumnName: "sName", FieldType: "string"},
+					"id":   {ColumnName: "ID", FieldType: "int64"},
+				},
+				defaultRules: "-name,id",
+			},
+			wantAPIError: ErrInvalidRequest(errors.New(`all 'from' parameters (from.name, from.id) or none of them must be present`))},
```

Comment:
Either the message and the logic are not really exact. (the message I think) You want all (or none) "from" parameters among the given sort attributes, right? If so, the message should say that.

Reply:
Could you add (or update this one) also a test where the expected parameters for "from" are not exactly the same as the fields accepted for sorting? (in short: add a field in "acceptedFields" which is not used)

Reply:
> Either the message and the logic are not really exact. (the message I think) You want all (or none) "from" parameters among the given sort attributes, right? If so, the message should say that.

The sort attributes are not always given in URL since the 'sort' parameter is optional. So the sort attributes can be default ones of a service as well.

Reply:
> Could you add (or update this one) also a test where the expected parameters for "from" are not exactly the same as the fields accepted for sorting? (in short: add a field in "acceptedFields" which is not used)

done

Reply:
> > Either the message and the logic are not really exact. (the message I think) You want all (or none) "from" parameters among the given sort attributes, right? If so, the message should say that.
> 
> The sort attributes are not always given in URL since the 'sort' parameter is optional. So the sort attributes can be default ones of a service as well.

My point was that each "from" parameters **must** be among the sorted fields (implicit or explicit). Do you validate that? I do not see an error message matching this.

In the same way, these "from" parameters **must** be fiven in the same order and the "sort" (implicit or explicit). Do you check that?

Reply:
We silently ignore all "from.*" parameters that are not among the sorted fields. It's consistent with the way we process usual get-parameters: all unknown parameters are just ignored.

"from.*" parameters are explored in the order of fields in the "sort" configuration. So there is no need to fix their order in URLs

Reply:
fixed